### PR TITLE
FOPTS-182 Kubecost request sizing recommendations policy template

### PR DESCRIPTION
**Pull request by vadgun**
### Description

The Kubecost Request Sizing Recommendations Policy Template is used to get recommendations for resizing instances controlled with Kubecost

### Issues Resolved
[Jira Ticket FOPTS-182](https://flexera.atlassian.net/browse/FOPTS-182)

### Links
Applied policy
[Applied policy on Flexera test](https://app.flexera.com/orgs/1105/automation/applied-policies/projects/105882?policyId=6452bce67b2db10001f8553d)

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 1]
```diff
@@ -0,0 +1,257 @@
+name "Kubecost Request Sizing Recommendations"
```
Comment (juanestrahcl) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:1]: Same as comment [here](https://github.com/flexera-public/policy_templates/pull/1237/files#diff-e150095efd6533a36cf0554ba0816ebee4c0aca0ebb159539f01a18ca0e9b62dR1).

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+datasource "ds_recommendations" do
+  run_script $js_requests_array, $ds_request_sizing
+end
+
+script "js_requests_array", type: "javascript" do
+  parameters "ds_request_sizing"
+  result "result"
+  code <<-EOS
+
```
Comment (juanestrahcl) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:150]: Remove this blank line.

File: cost/kubecost/sizing/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Kubecost Request Sizing Recommendations Policy Template
+
+## What it does
+
+The Kubecost Request Sizing Recommendations Policy Template is used to get recommendations for resizing instances controlled with Kubecost
```
Comment (degarajesh) [inline cost/kubecost/sizing/README.md:5]: Can we change this to ?
Kubecost request sizing recommendations policy template is used to resize instances controlled by Kubecost.

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+###############################################################################
+# Policy
+###############################################################################
+
+policy "policy_kubecost_request_sizing" do
+  validate $ds_recommendations do
+    summary_template "Kubecost Request Sizing Report"
+    detail_template <<-EOS
+The total estimated monthly savings : {{data.savings}}
+EOS
```
Comment (juanestrahcl) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:205]: Fix indentation of lines 204 and 205, should be at same level of 203.

File: cost/kubecost/sizing/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+## Functional Details
+
+- The policy receives recommendations to resize instances managed by Kubecost
+- The policy calls Kubecost API endpoint
+
+### Input Parameters
+
+- *Email addresses of the recipients you wish to notify* - A list of email addresses to notify
+- *Kubecost host* - The host or IP Address where your Kubecost runs
```
Comment (degarajesh) [inline cost/kubecost/sizing/README.md:24]: Kubecost Host or IP address

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+parameter "param_email" do
+  label "Email addresses"
+  description "Email addresses of the recipients you wish to notify"
+  type "list"
+end
+
+parameter "param_kubecost_host" do
+  type "string"
+  label "Kubecost Host"
+  description "IP Address of Kubecost LB"
```
Comment (degarajesh) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:30]: Kubecost Host or IP address

**Review by juanestrahcl — COMMENTED** [review]
Aside of comments I think you could add a link to an applied policy.

File: cost/kubecost/sizing/README.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# Kubecost Request Sizing Recommendations Policy Template
+
+## What it does
+
+The Kubecost Request Sizing Recommendations Policy Template is used to get recommendations for resizing instances controlled with Kubecost
+
+## Prerequisites
+
+This Policy Template uses [Credentials](https://docs.flexera.com/flexera/EN/Automation/ManagingCredentialsExternal.htm) for authenticating to datasources -- in order to apply this policy you must have a Credential registered in the system that is compatible with this policy. If there are no Credentials listed when you apply the policy, please contact your Flexera Org Admin and ask them to register a Credential that is compatible with this policy. The information below should be consulted when creating the credential(s).
```
Comment (degarajesh) [inline cost/kubecost/sizing/README.md:9]: Can we change this "To apply this policy, you must have a Credential registered in the system that is compatible with this policy..Please contact your Flexera Org Admin if there are no Credentials listed when you apply the policy and request that they register a Credential compatible with it.Please refer to the information below when creating the credential(s)."

**Comment by degarajesh** [comment]
And also can you please share the applied policies link ?

**Comment by vadgun** [comment]
> And also can you please share the applied policies link ?

Link To Apply Policy

[https://app.flexera.com/orgs/28010/automation/applied-policies/projects/123559?policyId=6434779f94f00300015498cb](url)

Thank you

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+rs_pt_ver 20180301
+type "policy"
+short_description "Pulls the Request Sizing Recommendations from Kubecost and raises and incident. See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/kubecost/sizing/) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
+long_description ""
+severity "low"
+category "Cost"
+default_frequency "daily"
+tenancy "single"
+info(
+  version: "1.0",
```
Comment (Rangelmv23) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:11]: Change version to "0.1"

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+end
+
+script "js_request_sizing", type: "javascript" do
+  parameters "kubecost_host", "param_cpu_algo", "param_cpu_q", "param_mem_algo", "param_mem_q", "param_window", "param_cpu_util", "param_mem_util"
+  result "request"
+  code <<-EOS
+  var request = {
+      "auth": "auth_kc",
+      "verb": "GET",
+      "scheme": "http",
```
Comment (degarajesh) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:135]: Please test with and without credentials

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+  type "string"
+  label "Range to analyze"
+  description "Accepts duration in number of days, hours or minutes e.g., 2d, 30m, 6h"
+  default "2d"
+end
+
+parameter "param_cpu_algo" do
+  type "string"
+  label "The algorithm for CPU"
+  description "The algorithm to be used to calculate CPU recommendations"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:43]: Defines whether max or a certain quantile should be used  to compute CPU request recommendation

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+  description "Target CPU utilization in percent"
+  min_value 1
+  max_value 100
+  default 80
+end
+
+parameter "param_mem_algo" do
+  type "string"
+  label "The algorithm for Memory"
+  description "The algorithm to be used to calculate memory recommendations"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:69]: Defines whether max or a certain quantile should be used  to compute Memory request recommendation

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+  label "The algorithm for CPU"
+  description "The algorithm to be used to calculate CPU recommendations"
+  allowed_values "max", "quantile"
+  default "max"
+end
+
+parameter "param_cpu_q" do
+  type "number"
+  label "CPU quantile"
+  description "The desired quantile to base CPU recommendations on"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:51]: The desired quantile to base CPU request recommendations on

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+  label "The algorithm for Memory"
+  description "The algorithm to be used to calculate memory recommendations"
+  allowed_values "max", "quantile"
+  default "max"
+end
+
+parameter "param_mem_q" do
+  type "number"
+  label "Memory quantile"
+  description "The desired quantile to base memory recommendations on"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:77]: The desired quantile to base Memory recommendations on

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+      field "savings" do
+        label "Savings"
+        format "center"
+      end
+      field "latestKnownRequestCPU" do
+        label "CPU Request"
+        format "center"
+      end
+      field "recommendedRequestCPU" do
+        label "REcommended CPU Request"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:225]: R**e**commended CPU Request

File: cost/kubecost/sizing/kubecost_resizing_recommendation.pt [hunk 15]
```diff
@@ -0,0 +1,249 @@
+name "Kubecost Request Right Sizing Recommendations"
```
Comment (TDubovchenko) [inline cost/kubecost/sizing/kubecost_resizing_recommendation.pt:1]: Let's replace "Right Sizing" with "Rightsizing"

**Review by bryankaraffa — APPROVED** [review]

**Pull request merged**